### DISCUSSION TITLE
Treat warning as error

### DISCUSSION THREAD

**Discussion by mdczaplicki** [sqlalchemy/sqlalchemy#6675] [Usage Questions] [answered]
In our project we sometimes forget to join two tables and it results with cartesian product of two tables. This kills our database.

I can see that SQLAlchemy is raising a warning about it:

```
sys:1: SAWarning: SELECT statement has a cartesian product between FROM element(s) "table_!" and FROM element "table_2".  Apply join condition(s) between each element to resolve.
```

Is there a way to make this warning an error? I'd like my tests to fail when this warning pops up.

**Comment by zzzeek** [comment, marked as answer]
Warnings can be errors using a variety of techniques surrounding the [warnings filter](https://docs.python.org/3/library/warnings.html).

A targeted approach here might be to put in your application:

```
   from sqlalchemy import exc
   import warnings

   warnings.filterwarnings("error", category=exc.SAWarning)
```

All warnings from SQLAlchemy will then be raised as errors.   you can tailor it to be more specific using a regex filter.

Also test suites should generally raise all warnings.  if you are using pytest for example, you'd want to set up the .ini to raise warnings as errors: https://docs.pytest.org/en/6.2.x/warnings.html

**Reply by mdczaplicki** [reply]
Thanks, first I wrote this:

```
@pytest.fixture(autouse=True)
def catch_warnings_as_exceptions(mocker: MockFixture):
    warnings_to_raise: List[Tuple[Type[Warning], str]] = [
        (SAWarning, r".*(cartesian)\s(product).*"),
    ]

    _execute = DatabaseAdapter._execute

    async def _execute_mock(self: DatabaseAdapter, query: Executable):
        with warnings.catch_warnings():
            for warning, regex in warnings_to_raise:
                warnings.filterwarnings("error", category=warning, message=regex)
            return await _execute(self, query)

    mocker.patch.object(DatabaseAdapter, "_execute", _execute_mock)
```

but now I've refactored it to this in `pytest.ini` (after your comment):

```
filterwarnings =
    error:.*(cartesian)\s(product).*:sqlalchemy.exc.SAWarning
```
